validator/sawtooth_validator/networking/interconnect.py

```diff
@@ -230,7 +230,9 @@ def _do_heartbeat(self):
                                          " - removing connection.",
                                          self._connection,
                                          self._connection_timeout)
-                            self._ready.clear()
+                            connection_id = hashlib.sha512(
+                                self.connection.encode()).hexdigest()
+                            del self._connections[connection_id]
```

Comment:
I would guard this with a `if connection_id in self._connections:` ; we have had issues where this collection get's modified by multiple threads